ml_perf/reference_implementation.py

```diff
@@ -0,0 +1,279 @@
+from absl import app, flags
+from rl_loop import example_buffer, fsdb, shipname
+
+flags.DEFINE_string('engine', 'tf', 'Engine to use for inference.')
+
+FLAGS = flags.FLAGS
+
+
+class State:
+
+  _NAMES = ['bootstrap'] + random.Random(0).sample(shipname.NAMES,
+                                                   len(shipname.NAMES))
+
+  def __init__(self):
+    self.iter_num = 0
+    self.play_model_num = 0
+    self.play_model_name = self.play_output_name
+    self.train_model_num = 1
+
+  @property
+  def play_output_name(self):
+    return '%06d-%s' % (self.iter_num, self._NAMES[self.play_model_num])
+
+  @property
+  def play_model_path(self):
+    return os.path.join(fsdb.models_dir(), self.play_model_name)
+
+  @property
+  def train_model_name(self):
+    return '%06d-%s' % (self.iter_num, self._NAMES[self.train_model_num])
+
+  @property
+  def train_model_path(self):
+    return os.path.join(fsdb.models_dir(), self.train_model_name)
+
+  @property
+  def seed(self):
+    return self.iter_num + 1
+
+
+def checked_run(cmd, name):
+  logging.info('Running %s:\n  %s', name, '\n  '.join(cmd))
+  with utils.logged_timer('%s finished' % name.capitalize()):
+    completed_process = subprocess.run(
+        cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    if completed_process.returncode:
+      logging.error('Error running %s: %s', name,
+                    completed_process.stdout.decode())
+      raise RuntimeError('Non-zero return code executing %s' % ' '.join(cmd))
+  return completed_process
+
+
+def get_lines(completed_process, slice):
+  return '\n'.join(completed_process.stdout.decode()[:-1].split('\n')[slice])
+
+
+class MakeSlice(object):
+
+  def __getitem__(self, item):
+    return item
+
+
+make_slice = MakeSlice()
+
+
+def cc_flags(state):
+  return [
+      '--engine={}'.format(FLAGS.engine),
+      '--virtual_losses=8',
+      '--seed={}'.format(state.seed),
+  ]
+
+
+def py_flags(state):
+  return [
+      '--work_dir={}'.format(fsdb.working_dir()),
+      '--trunk_layers=10',
+      '--conv_width=64',
+      '--value_cost_weight=0.25',
+      '--training_seed={}'.format(state.seed),
+  ]
+
+
+# Generate an initial model with random weights.
+def bootstrap(state):
+  checked_run([
+      'python3', 'bootstrap.py', '--export_path={}'.format(
+          state.play_model_path)
+  ] + py_flags(state), 'bootstrap')
+
+
+# Freezes a model checkpoint into a TensorFlow GraphDef proto, which is the
+# required model format for C++ TensorFlow.
+def freeze(state, model_path):
+  checked_run([
+      'python3', 'freeze_graph.py', '--model_path={}'.format(model_path)
+  ] + py_flags(state), 'freeze')
+
+
+# Self-play a number of games.
+def selfplay(state):
+  play_output_name = state.play_output_name
+  play_output_dir = os.path.join(fsdb.selfplay_dir(), play_output_name)
+  play_holdout_dir = os.path.join(fsdb.holdout_dir(), play_output_name)
+
+  result = checked_run([
+      'bazel-bin/cc/selfplay', '--parallel_games=2048',
```

Comment:
that's a lot of parallel games :D
I guess low readouts means you can support them all. Seems like too few readouts to make this work properly, but I guess that's up to MLPerf guys to optimize.

Reply:
yeah, Christian was running 9x9 on 8 v100s, which required a large number of parallel games to get decent utilization, as you might imagine.

We will want to support playing N games with M<N in parallel, so that games will finish quicker for 19x19 and maybe also help with the ragged edge problem.